Mathlib/MeasureTheory/Decomposition/Exhaustion.lean

```diff
@@ -0,0 +1,335 @@
+
+lemma measurableSet_sigmaFiniteSetWRT' (μ ν : Measure α) [IsFiniteMeasure ν] :
+    MeasurableSet (μ.sigmaFiniteSetWRT' ν) :=
+  MeasurableSet.iUnion (measurableSet_sigmaFiniteSetGE _ _)
+
+instance [IsFiniteMeasure ν] : SigmaFinite (μ.restrict (μ.sigmaFiniteSetWRT' ν)) := by
+  let f : ℕ × ℕ → Set α := fun p : ℕ × ℕ ↦ (μ.sigmaFiniteSetWRT' ν)ᶜ
+    ∪ (spanningSets (μ.restrict (μ.sigmaFiniteSetGE ν p.1)) p.2 ∩ (μ.sigmaFiniteSetGE ν p.1))
+  suffices (μ.restrict (μ.sigmaFiniteSetWRT' ν)).FiniteSpanningSetsIn (Set.range f) from
+    this.sigmaFinite
+  let e : ℕ ≃ ℕ × ℕ := Nat.pairEquiv.symm
+  refine ⟨fun n ↦ f (e n), fun _ ↦ by simp, fun n ↦ ?_, ?_⟩
+  · simp only [Nat.pairEquiv_symm_apply, gt_iff_lt, measure_union_lt_top_iff, f, e]
+    rw [Measure.restrict_apply' (measurableSet_sigmaFiniteSetWRT' _ _), Set.compl_inter_self,
+      Measure.restrict_apply' (measurableSet_sigmaFiniteSetWRT' _ _)]
+    simp only [measure_empty, ENNReal.zero_lt_top, true_and]
+    refine (measure_mono Set.inter_subset_left).trans_lt ?_
+    rw [← Measure.restrict_apply' (measurableSet_sigmaFiniteSetGE μ ν _)]
+    exact measure_spanningSets_lt_top _ _
+  · simp only [Nat.pairEquiv_symm_apply, f, e]
+    rw [← Set.union_iUnion]
+    suffices ⋃ n, (spanningSets (μ.restrict (μ.sigmaFiniteSetGE ν (Nat.unpair n).1)) n.unpair.2
+        ∩ μ.sigmaFiniteSetGE ν n.unpair.1) = μ.sigmaFiniteSetWRT' ν by
+      rw [this, Set.compl_union_self]
+    calc ⋃ n, (spanningSets (μ.restrict (μ.sigmaFiniteSetGE ν (Nat.unpair n).1)) n.unpair.2
+        ∩ μ.sigmaFiniteSetGE ν n.unpair.1)
+      = ⋃ n, ⋃ m, (spanningSets (μ.restrict (μ.sigmaFiniteSetGE ν n)) m
+            ∩ μ.sigmaFiniteSetGE ν n) :=
+          Set.iUnion_unpair (fun n m ↦ spanningSets (μ.restrict (μ.sigmaFiniteSetGE ν n)) m
+            ∩ μ.sigmaFiniteSetGE ν n)
+    _ = ⋃ n, μ.sigmaFiniteSetGE ν n := by
+        refine Set.iUnion_congr (fun n ↦ ?_)
+        rw [← Set.iUnion_inter, iUnion_spanningSets, Set.univ_inter]
+    _ = μ.sigmaFiniteSetWRT' ν := rfl
+
+/-- `μ.sigmaFiniteSetWRT' ν` has maximal `ν`-measure among all measurable sets `s` with sigma-finite
+`μ.restrict s`. -/
+lemma measure_sigmaFiniteSetWRT' (μ ν : Measure α) [IsFiniteMeasure ν] :
+    ν (μ.sigmaFiniteSetWRT' ν)
+      = ⨆ (s) (_ : MeasurableSet s) (_ : SigmaFinite (μ.restrict s)), ν s := by
+  apply le_antisymm
+  · refine (le_iSup (f := fun _ ↦ _)
+      (inferInstance : SigmaFinite (μ.restrict (μ.sigmaFiniteSetWRT' ν)))).trans ?_
+    exact le_iSup₂ (f := fun s _ ↦ ⨆ (_ : SigmaFinite (μ.restrict s)), ν s) (μ.sigmaFiniteSetWRT' ν)
+      (measurableSet_sigmaFiniteSetWRT' μ ν)
+  · exact le_of_tendsto' (tendsto_measure_sigmaFiniteSetGE μ ν)
+      (fun _ ↦ measure_mono (Set.subset_iUnion _ _))
+
+/-- For all measurable sets `s` in `(μ.sigmaFiniteSetWRT ν)ᶜ`, if `ν s ≠ 0` then `μ s = ∞`. -/
+lemma measure_eq_top_of_subset_compl_sigmaFiniteSetWRT'
+    (hs : MeasurableSet s) (hs_subset : s ⊆ (μ.sigmaFiniteSetWRT' ν)ᶜ) (hνs : ν s ≠ 0) :
```

Comment:
Once you've proved this one, can you drop the `MeasurableSet` assumption, by arguing that the measure is the inf of the measure of measurable sets containing them (which you can take inside `(μ.sigmaFiniteSetWRT' ν)ᶜ` as it's measurable)?